src/util/Channel.hpp

```diff
@@ -0,0 +1,195 @@
+        {
+        }
+
+        [[nodiscard]] auto&
+        channel()
+        {
+            return ch_;
+        }
+
+        void
+        close()
+        {
+            ch_.close();
+        }
+
+        [[nodiscard]] bool
+        isClosed() const
+        {
+            return not ch_.is_open();
+        }
+    };
+
+    class Sender {
+        std::shared_ptr<Shared> shared_;
+        struct Inner {
+            std::shared_ptr<Shared> shared;
+
+            ~Inner()
+            {
+                shared->close();
+            }
+        };
+        std::shared_ptr<Inner> inner_;
+
+    public:
+        Sender(std::shared_ptr<Shared> shared)
+            : shared_(std::move(shared)), inner_(std::make_shared<Inner>(shared_)) {};
+        Sender(Sender&&) = default;
+        Sender(Sender const&) = default;
+        Sender&
+        operator=(Sender&&) = default;
+        Sender&
+        operator=(Sender const&) = default;
+
+        template <typename D>
+        bool
+        asyncSend(D&& data, boost::asio::yield_context yield)
+            requires(std::same_as<std::remove_cvref_t<T>, std::remove_cvref_t<D>>)
+        {
+            boost::system::error_code ec;
+            shared_->channel().async_send(ec, std::forward<decltype(data)>(data), yield[ec]);
+            return not ec;
+        }
+
+        template <typename D>
+        void
+        asyncSend(D&& data, std::invocable<bool> auto&& fn)
+            requires(std::same_as<std::remove_cvref_t<T>, std::remove_cvref_t<D>>)
+        {
+            boost::system::error_code ec;
+            shared_->channel().async_send(
+                ec,
+                std::forward<decltype(data)>(data),
+                [fn = std::forward<decltype(fn)>(fn)](boost::system::error_code ec) mutable { fn(not ec); }
+            );
+        }
+
+        template <typename D>
+        bool
+        trySend(D&& data)
+            requires(std::same_as<std::remove_cvref_t<T>, std::remove_cvref_t<D>>)
+        {
+            boost::system::error_code ec;
+            return shared_->channel().try_send(ec, std::forward<decltype(data)>(data));
+        }
+    };
+
+    class Receiver {
+        std::shared_ptr<Shared> shared_;
+
+    public:
+        Receiver(std::shared_ptr<Shared> shared) : shared_(std::move(shared)) {};
+        Receiver(Receiver&&) = default;
+        Receiver(Receiver const&) = delete;
+        Receiver&
+        operator=(Receiver&&) = default;
+        Receiver&
+        operator=(Receiver const&) = delete;
+
+        std::optional<T>
+        tryReceive()
+        {
+            std::optional<T> result;
+            auto const received = shared_->channel().try_receive([&result](boost::system::error_code ec, auto&& value) {
+                if (not ec)
+                    result = std::forward<decltype(value)>(value);
+            });
+
+            if (received)
```

Comment:
Is it possible for some reason to have error in `ec`, but received True?
And why is there 2 ways to communicate the error?